Development/nmos/constraints.cpp

```diff
@@ -0,0 +1,284 @@
+            }
+            else if (rhs.has_field(nmos::fields::constraint_minimum))
+            {
+                result[nmos::fields::constraint_minimum] = nmos::fields::constraint_minimum(rhs);
+            }
+
+            // "maximum"
+            if (lhs.has_field(nmos::fields::constraint_maximum) && rhs.has_field(nmos::fields::constraint_maximum))
+            {
+                result[nmos::fields::constraint_maximum] = std::min(nmos::fields::constraint_maximum(lhs), nmos::fields::constraint_maximum(rhs), constraint_comp);
+            }
+            else if (lhs.has_field(nmos::fields::constraint_maximum))
+            {
+                result[nmos::fields::constraint_maximum] = nmos::fields::constraint_maximum(lhs);
+            }
+            else if (rhs.has_field(nmos::fields::constraint_maximum))
+            {
+                result[nmos::fields::constraint_maximum] = nmos::fields::constraint_maximum(rhs);
+            }
+
+            // "min" > "max"
+            if (result.has_field(nmos::fields::constraint_minimum) && result.has_field(nmos::fields::constraint_maximum))
+            {
+                if (!constraint_comp(nmos::fields::constraint_minimum(result), nmos::fields::constraint_maximum(result)))
+                {
+                    return web::json::value::null();
+                }
+            }
+
+            // "enum" matches "min"/"max"
+            if (result.has_field(nmos::fields::constraint_enum) && (result.has_field(nmos::fields::constraint_minimum) || result.has_field(nmos::fields::constraint_maximum)))
+            {
+                result[nmos::fields::constraint_enum] = get_intersection(nmos::fields::constraint_enum(result).as_array(), result);
+                for (const auto& keyword : { nmos::fields::constraint_minimum, nmos::fields::constraint_maximum })
+                {
+                    if (result.has_field(keyword))
+                    {
+                        result.erase(keyword);
+                    }
+                }
+            }
+
+            // no "min"/"max" and "enum" is empty
+            if (result.has_field(nmos::fields::constraint_enum) && (!result.has_field(nmos::fields::constraint_minimum) && !result.has_field(nmos::fields::constraint_maximum)))
+            {
+                if (0 == nmos::fields::constraint_enum(result).as_array().size())
+                {
+                    return web::json::value::null();
+                }
+            }
+
+            return result;
+        }
+
+        web::json::value get_constraint_set_intersection(const web::json::value& lhs, const web::json::value& rhs, bool merge_left_to_right)
+        {
+            using web::json::value;
+            using web::json::value_from_elements;
+
+            const auto& lhs_param_constraints = lhs.as_object();
+            auto rhs_param_constraints = rhs.as_object();
+
+            if (merge_left_to_right)
```

Comment:
I may be completely wrong here, but my logic was that calculating the intersection is picking common items from two sets into a new set. Thus, if we consider two objects with different sets of keys, the intersection should have only the common keys by default. Does it make sense?

Reply:
We discussed so just noting for other readers, the answer is no, because parameter constraints are and-ed together so the intersection of two constraint sets needs all the parameter constraints from both sets. This is different than calculating the intersection for the arrays of values under the enum constraint keyword because those values are or-ed.